library/src-bootstrapped/scala/reflect/TypeTest.scala

```diff
@@ -0,0 +1,28 @@
+package scala.reflect
+
+/** A `TypeTest[S, T] contains the logic needed to know at runtime if a value of
+ *  type `S` can be downcasted to `T`.
+ *
+ *  If a pattern match is performed on a term of type `s: S` that is uncheckable with `s.isInstanceOf[T]` and
+ *  the pattern are of the form:
+ *    - `t: T`
+ *    - `t @ X()` where the `X.unapply` has takes an argument of type `T`
+ *  then a given instance of `TypeTest[S, T]` is summoned and used to perform the test.
+ */
+@scala.annotation.implicitNotFound(msg = "No TypeTest available for [${S}, ${T}]")
+trait TypeTest[-S <: Matchable, T <: Matchable] extends Serializable:
```

Comment:
These restrictions at this level are incorrect. We must not constrain the type parameters in `TypeTest`. Otherwise it defeats the whole purpose of `TypeTest`. For example the whole TASTy-Reflect API relies on having `S` and `T`s that are *not* `<: Matchable`.

The only thing we should restrict is *synthesizing* `TypeTest` with an `S` that is not `Matchable` (because the synthesized code calls `isInstanceOf`/performs a `match` on the input of type `S`.

Reply:
https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/typer/Synthesizer.scala#L65